RIOT-OS/RIOT
# boards: make: introduce mechanism to define supported programmers

### Contribution description

This PR introduces basic mechanism to specify, at board level, the list of supported programmers. List of supported programmers is set in the `PROGRAMMERS_SUPPORTED` variables.
A check in the main Makefile.include is added to verify that the requested programmer is supported by the board. If it's not, the build fails with a meaningful error message. If `PROGRAMMERS_SUPPORTED` is undefined all programmers are accepted.

For most of the boards, `PROGRAMMERS_SUPPORTED` is set in a central place to reduce code duplication. Other mechanisms are used with nrf boards to set blacklisted programmers or extra programmers at board level if the common list doesn't match.

Most of the groups of boards are adapted in this PR: nucleo, stm32 with st-link, sam0 using a bootloader, nrf, frdm, silabs, atmega.

I also noticed that sam0.inc.mk include in the itsybitsy, serpente and yarm boards was useless as it's content is mainly about edbg, openocd, jlink while these boards are flashed using bossa.

There's also a new target `info-programmers-supported` added to print the content of `PROGRAMMERS_SUPPORTED` for a given board.

Also making this as RFC because this is an important change (lots of boards touched).




### Testing procedure

- A green Murdock will verify that there's no miss in the changes
- Of course, verify that current "supported" programmers can still be used and works. I don't have a precise procedure for that other than take each affected boards and try all programmers on it.

Here is a set of examples with a few boards:

<details><summary>nucleo: only openocd is supported</summary>

```
$ make BOARD=nucleo-l073rz -C examples/hello-world info-programmers-supported --no-print-directory 
openocd
```

</details>

<details><summary>nucleo: jlink is not supported</summary>

```
$ PROGRAMMER=jlink make BOARD=nucleo-l073rz -C examples/hello-world --no-print-directory 
/work/riot/RIOT/examples/hello-world/../../Makefile.include:416: *** 'jlink' programmer is not supported by this board. Supported programmers: 'openocd'.  Stop.
```

</details>

<details><summary>samr21-xpro: several programmers are supported (openocd, jlink, edbg</summary>

```
$ make BOARD=samr21-xpro -C examples/hello-world info-programmers-supported --no-print-directory 
openocd edbg jlink
```

</details>

<details><summary>hifive1b: jlink and openocd are supported</summary>

```
$ make BOARD=hifive1b -C examples/hello-world info-programmers-supported --no-print-directory 
jlink openocd
```

</details>

<details><summary>arduino-nano-33-iot: only bossa is supported</summary>

```
$ make BOARD=arduino-nano-33-iot -C examples/hello-world info-programmers-supported --no-print-directory 
bossa
```

</details>

<details><summary>arduino-mkr1000: bossa and jlink are supported</summary>

```
$ make BOARD=arduino-mkr1000 -C examples/hello-world info-programmers-supported --no-print-directory 
bossa jlink
```

</details>

<details><summary>nrf52dk: bmp, openocd and jlink are supported</summary>

```
$ make BOARD=nrf52dk -C examples/hello-world info-programmers-supported --no-print-directory 
bmp jlink openocd
```

</details>

<details><summary>nrf52832-mdk: bmp, openocd and pyocd are supported</summary>

```
$ make BOARD=nrf52832-mdk -C examples/hello-world info-programmers-supported --no-print-directory 
bmp openocd pyocd
```

</details>

<details><summary>microbit: jlink, openocd and pyocd are supported</summary>

```
$ make BOARD=microbit -C examples/hello-world info-programmers-supported --no-print-directory 
jlink openocd pyocd
```

</details>

<details><summary>atmega256rfr2-xpro: only avrdude is supported</summary>

```
$ make BOARD=atmega256rfr2-xpro -C examples/hello-world info-programmers-supported --no-print-directory 
avrdude
```

</details>

<details><summary>atmega256rfr2-xpro: openocd is not supported</summary>

```
$ PROGRAMMER=openocd make BOARD=atmega256rfr2-xpro -C examples/hello-world --no-print-directory /work/riot/RIOT/examples/hello-world/../../Makefile.include:416: *** 'openocd' programmer is not supported by this board. Supported programmers: 'avrdude'.  Stop.
```

</details>




### Issues/PRs references

None
